### ISSUE TITLE
Problems with Instruction API parsing x86-64 binaries: xhpl executable

### ISSUE THREAD

**Issue by jmellorcrummey** [dyninst/dyninst#80]
The current master version of dyninst has problems parsing several instructions in x86-64 binaries. These examples of problems were identified using the checkparse regression test developed at Rice and provided to Wisconsin.

**Application example: xhpl**

The x86-64 xhpl binary may be found here: http://www.cs.rice.edu/~johnmc/tmp/xhpl

problematic instructions:
```
            pop    (xed length=2, instruction api length=1) last instruction in block with DIRECT edge to sink
          kmovw    instruction length mismatch (xed length=8, instruction api length=4)
          kmovw    instruction length mismatch (xed length=9, instruction api length=4)
          vdppd    (xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
         pinsrq    (xed length=12, instruction api length=5) last instruction in block with DIRECT edge to sink
         psrldq    (xed length=6, instruction api length=2) last instruction in block with DIRECT edge to sink
         vpsrlq    (xed length=11, instruction api length=5) last instruction in block with DIRECT edge to sink
         vpsrlq    (xed length=7, instruction api length=5) last instruction in block with DIRECT edge to sink
        vmovdqa    (xed length=5, instruction api length=4) last instruction in block with DIRECT edge to sink
        vpermpd    (xed length=7, instruction api length=5) last instruction in block with DIRECT edge to sink
       vblendpd    (xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
       vfmaddpd    (xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
     vgatherqpd    (xed length=10, instruction api length=4) last instruction in block with DIRECT edge to sink
   vpbroadcastd    (xed length=6, instruction api length=5) last instruction in block with DIRECT edge to sink
   vpbroadcastq    (xed length=6, instruction api length=5) last instruction in block with DIRECT edge to sink
```

Attached is the full report from checkparse. Unfortunately, I don't have the time to extract reports for issues for individual instruction instances. The checkparse output identifies problematic occurences. For addresses where xed claims no valid instruction at this address, the problem may be  unparsing a prior instruction in a block rather than the last one. 

[xhpl.report.txt](https://github.com/dyninst/dyninst/files/305946/xhpl.report.txt)

**Comment by mxz297** [comment]
My preliminary investigation found two issues:

1. Jump table analysis was over-approximating the size of a jump table, which led to one of the many "xed decode error;no instruction at address" reports. It appears to me that our x86-64 instruction semantics missed semantics for "cmove" and in this case understanding "cmove" correctly is essentail to precisely determine the table size. I will continue to fix this jump table analysis problem.

2. My mutator hit an assertation failure when decoding instruction at address 0x4697cc, which should be from a valid instruction stream as several instructions before this address align with the output from objdump. This problem does not seem to be related to the error reporting from Rice, but is still worth further investigation.

The problematic instruction is 

` 4697cc:       c4 e2 39 f7 d3          shlx   %r8d,%ebx,%edx`

The stack trace is below:
```
CodeDump: /p/paradyn/development/xmeng/dyninstapi/dyninst/instructionAPI/src/InstructionDecoder-x86.C:532: Dyninst::MachRegister Dyninst::InstructionAPI::IntelRegTable_access::operator()(Dyninst::Architecture, Dyninst::InstructionAPI::intelRegBanks, int): Assertion `index >= 0 && index < 8' failed.

Program received signal SIGABRT, Aborted.
0x00007ffff5bab5d7 in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install boost-system-1.53.0-23.el7.x86_64 boost-thread-1.53.0-23.el7.x86_64 elfutils-libelf-0.160-1.el7.x86_64 glibc-2.17-78.el7.x86_64 libgcc-4.8.3-9.el7.x86_64 libstdc++-4.8.3-9.el7.x86_64
(gdb) bt
#0  0x00007ffff5bab5d7 in raise () from /lib64/libc.so.6
#1  0x00007ffff5baccc8 in abort () from /lib64/libc.so.6
#2  0x00007ffff5ba4546 in __assert_fail_base () from /lib64/libc.so.6
#3  0x00007ffff5ba45f2 in __assert_fail () from /lib64/libc.so.6
#4  0x00007ffff765cd58 in Dyninst::InstructionAPI::IntelRegTable_access::operator() (arch=<optimized out>, bank=bank@entry=Dyninst::InstructionAPI::b_32bit, index=index@entry=8, this=<optimized out>)
    at /p/paradyn/development/xmeng/dyninstapi/dyninst/instructionAPI/src/InstructionDecoder-x86.C:532
#5  0x00007ffff765d300 in Dyninst::InstructionAPI::InstructionDecoder_x86::makeRegisterID (this=this@entry=0x13e2650, intelReg=8, opType=opType@entry=16, isExtendedReg=<optimized out>)
    at /p/paradyn/development/xmeng/dyninstapi/dyninst/instructionAPI/src/InstructionDecoder-x86.C:651
#6  0x00007ffff765fc38 in Dyninst::InstructionAPI::InstructionDecoder_x86::decodeOneOperand (this=this@entry=0x13e2650, b=..., operand=..., imm_index=@0x7fffffffd2cc: 0, insn_to_complete=insn_to_complete@entry=0x952bfd0, 
    isRead=<optimized out>, isWritten=isWritten@entry=false) at /p/paradyn/development/xmeng/dyninstapi/dyninst/instructionAPI/src/InstructionDecoder-x86.C:988
#7  0x00007ffff76623a4 in Dyninst::InstructionAPI::InstructionDecoder_x86::decodeOperands (this=0x13e2650, insn_to_complete=0x952bfd0)
    at /p/paradyn/development/xmeng/dyninstapi/dyninst/instructionAPI/src/InstructionDecoder-x86.C:1787
#8  0x00007ffff7662d08 in Dyninst::InstructionAPI::InstructionDecoder_x86::doDelayedDecode (this=0x13e2650, insn_to_complete=0x952bfd0)
    at /p/paradyn/development/xmeng/dyninstapi/dyninst/instructionAPI/src/InstructionDecoder-x86.C:1829
#9  0x00007ffff76407f0 in Dyninst::InstructionAPI::Instruction::decodeOperands (this=this@entry=0x952bfd0) at /p/paradyn/development/xmeng/dyninstapi/dyninst/instructionAPI/src/Instruction.C:120
#10 0x00007ffff7640efe in Dyninst::InstructionAPI::Instruction::format (this=this@entry=0x952bfd0, addr=addr@entry=0) at /p/paradyn/development/xmeng/dyninstapi/dyninst/instructionAPI/src/Instruction.C:445
#11 0x0000000000401911 in DumpInstructionForBlock (f=f@entry=0xdd86a0, b=<optimized out>) at CodeDump.cpp:27
#12 0x000000000040152c in main (argc=<optimized out>, argv=<optimized out>) at CodeDump.cpp:64

```

@jdetter: Could you take a look at this three operands decoding issue? I think you can trigger this problem by calling the format method of this instruction.

**Comment by jdetter** [comment]
Yup, I'll look into it.

**Comment by jdetter** [comment]
I have fixed the assert issue, it was an issue with the r8-r15 64 bit registers being accessed from the VEX prefix. I am working on the instruction length fixes now.

**Comment by jdetter** [comment]
@jmellorcrummey I have uploaded some significant fixes for your binaries. Do you think you could regenerate the report to see what the remaining problems are? The report output was really helpful.

**Comment by jmellorcrummey** [comment]
I’ll do that tonight. I’ll also send you my new testing harness (off list).
--
John Mellor-Crummey         Professor
Dept of Computer Science    Rice University
email: johnmc@rice.edu      phone: 713-348-5179

**Comment by jmellorcrummey** [comment]
Are your fixes in the VEX branch or master?
--
John Mellor-Crummey         Professor
Dept of Computer Science    Rice University
email: johnmc@rice.edu      phone: 713-348-5179

**Comment by jdetter** [comment]
They are on the VEX branch. 

On my end there are still a significant amount of issues but much less than before. There's an issue with VEX instructions prefixed with a REX prefix. I should have that figured out soon.

**Comment by jmellorcrummey** [comment]
With your changes in the VEX branch, the situation is much improved for me for xhpl and several of the other test cases that I provided.

Error reports for troublesome binaries (measured by wc)

Before:
    75    522   9647 v1/hpgmg-fv.report.txt
     1      9    124 v1/lulesh-RAJA-parallel.exe.report.txt
     6    154   1710 v1/mhdshock.report.txt
     6    154   1710 v1/pulse.report.txt
     6    154   1710 v1/soliton.report.txt
    22    156   2587 v1/ucns3d_p.report.txt
  1383  21884 267547 v1/xhpl.report.txt
  1499  23033 285035 total

After:
   121   1360  18528 hpgmg-fv.report.txt
     1     15    155 lulesh-RAJA-parallel.exe.report.txt
     0      0      0 mhdshock.report.txt
     0      0      0 pulse.report.txt
     0      0      0 soliton.report.txt
    61    789   9529 ucns3d_p.report.txt
   330   6107  73100 xhpl.report.txt
   513   8271 101312 total

Note though for hpgmg-fv and ucns3d_p , the number of error reports went up. 

After: hpgmg-fv.mishandled-instructions.txt 
            jmp    DIRECT edge to sink
         pinsrb    DIRECT edge to sink
         psrldq    DIRECT edge to sink
         xbegin    DIRECT edge to sink
        vmovdqa    DIRECT edge to sink
       vblendps    DIRECT edge to sink

After: hpgmg-fv.mishandled-instructions.txt 
            jmp    (xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
         pinsrb    (xed length=7, instruction api length=4) last instruction in block with DIRECT edge to sink
         pinsrb    (xed length=8, instruction api length=5) last instruction in block with DIRECT edge to sink
         psrldq    (xed length=6, instruction api length=2) last instruction in block with DIRECT edge to sink
         vpsrlq    (xed length=5, instruction api length=3) last instruction in block with DIRECT edge to sink
         vpsrlq    (xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
         xbegin    (xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
        vmovdqa    (xed length=5, instruction api length=4) last instruction in block with DIRECT edge to sink
        vpsrldq    (xed length=5, instruction api length=3) last instruction in block with DIRECT edge to sink
        vpsrldq    (xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
       vblendps    (xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink

Before v1/ucns3d_p.mishandled-instructions.txt 
          vdppd    DIRECT edge to sink
        vmovdqa    DIRECT edge to sink

After ucns3d_p.mishandled-instructions.txt 
         vpsrlq    (xed length=5, instruction api length=3) last instruction in block with DIRECT edge to sink
        vmovdqa    (xed length=5, instruction api length=4) last instruction in block with DIRECT edge to sink
        vpsrldq    (xed length=5, instruction api length=3) last instruction in block with DIRECT edge to sink
        vpsrldq    (xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink

--
John Mellor-Crummey         Professor
Dept of Computer Science    Rice University
email: johnmc@rice.edu      phone: 713-348-5179

**Comment by jdetter** [comment]
Your binary exercised and issue in our table that was kind of complicated. It seems to have opened up a can of worms in our decoding tables. 

The bad news is that we were decoding a significant amount of instructions (non VEX) completely wrong, and potentially with buffer overreads. The good news is that the rows are very straight forward to fix and I should be able to make a ton of progress on this today. I will keep you updated!

**Comment by mxz297** [comment]
Updates on jump table analysis. I get the one I mentioned above fixed, but I found several other issues about jump table analysis. Basically, if I found overlapping instructions in an executable and if the overlapping instructions have incoming edges from jump tables, it is probably that the jump table analysis is doing something wrong. I think if I get these issues resolved, we should see much fewer error reports.

I have been making good progress and am still working with them.

**Comment by jdetter** [comment]
@jmellorcrummey the current master should contain a ton of instruction fixes. Could you run your report program again and post the results as a gist? The full report was really helpful for finding the locations of the problem instructions.

**Comment by jmellorcrummey** [comment]
Full issues with xhpl and test binaries provided are listed below. Things are much better:

```
    9   135   999 hpgmg-fv.mishandled-instructions.txt
   57   702  9679 hpgmg-fv.report.txt
    0     0     0 lulesh-RAJA-parallel.exe.hpcstruct.mishandled-instructions.txt
    0     0     0 lulesh-RAJA-parallel.exe.hpcstruct.report.txt
    0     0     0 lulesh-RAJA-parallel.exe.mishandled-instructions.txt
    0     0     0 lulesh-RAJA-parallel.exe.report.txt
    0     0     0 mhdshock.mishandled-instructions.txt
    0     0     0 mhdshock.report.txt
    0     0     0 pulse.mishandled-instructions.txt
    0     0     0 pulse.report.txt
    0     0     0 soliton.mishandled-instructions.txt
    0     0     0 soliton.report.txt
    1    15   111 ucns3d_p.mishandled-instructions.txt
   20   156  2242 ucns3d_p.report.txt
    2    30   222 xhpl.mishandled-instructions.txt
  104   651 12006 xhpl.report.txt
  193  1689 25259 total

```

```
hpgmg-fv.report.txt
fn=(__kmp_print_storage_map_gtid,0x693b40);bl=0x693b8e;addr=0x693bbb;;xed decode error;no instruction at address
fn=(__kmp_print_storage_map_gtid,0x693b40);bl=0x693b8e;addr=0x693bbb;;xed decode error;no instruction at address
fn=(__kmp_print_storage_map_gtid,0x693b40);bl=0x693b8f;addr=0x693bbb;;xed decode error;no instruction at address
fn=(__kmp_print_storage_map_gtid,0x693b40);bl=0x693b8f;addr=0x693bbb;;xed decode error;no instruction at address
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_test_adaptive_lock,0x728e30);bl=0x728e7a;addr=0x728e80;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_test_adaptive_lock_with_checks,0x728ee0);bl=0x728fa4;addr=0x728faa;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_acquire_adaptive_lock,0x729020);bl=0x729198;addr=0x72919e;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_acquire_adaptive_lock,0x729020);bl=0x7291f8;addr=0x7291fe;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_acquire_adaptive_lock_with_checks,0x729360);bl=0x7295cc;addr=0x7295d2;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_acquire_adaptive_lock_with_checks,0x729360);bl=0x729622;addr=0x729628;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_acquire_rtm_lock,0x72b080);bl=0x72b080;addr=0x72b09c;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_acquire_rtm_lock_with_checks,0x72b310);bl=0x72b310;addr=0x72b32c;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_test_rtm_lock,0x72b780);bl=0x72b780;addr=0x72b788;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(_INTERNAL_74__global_cscratch1_sd_laksono_pkgs_src_LLVM_openmp_runtime_src_kmp_lock_cpp_b0197b55::__kmp_test_rtm_lock_with_checks,0x72b7d0);bl=0x72b7d0;addr=0x72b7d8;xbegin;(6 bytes);(xed length=6, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(hp_timing_gettime,0x774b60);bl=0x774b8c;addr=0x774ba6;jmp;(5 bytes);(xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(MPL_cray_trdmapp_sheap_free,0x887420);bl=0x8874a1;addr=0x8874b0;jmp;(5 bytes);(xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(__svml_urem4_h9,0x9af520);bl=0x9af520;addr=0x9af551;vblendps;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__svml_urem4_y8,0x9af5f0);bl=0x9af5f0;addr=0x9af622;pblendw;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strspn,0xa4dde0);bl=0xa4de16;addr=0xa4de16;;xed decode error;no instruction at address
fn=(__intel_sse4_strspn,0xa4dde0);bl=0xa4de40;addr=0xa4de40;;xed decode error;no instruction at address
fn=(__intel_sse4_strspn,0xa4dde0);bl=0xa4de6a;addr=0xa4de6a;;xed decode error;no instruction at address
fn=(__intel_sse4_strspn,0xa4dde0);bl=0xa4e048;addr=0xa4e048;pcmpistrm;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strcspn,0xa4e110);bl=0xa4e166;addr=0xa4e166;;xed decode error;no instruction at address
fn=(__intel_sse4_strcspn,0xa4e110);bl=0xa4e16d;addr=0xa4e16d;;xed decode error;no instruction at address
fn=(__intel_sse4_strcspn,0xa4e110);bl=0xa4e174;addr=0xa4e174;;xed decode error;no instruction at address
fn=(__intel_sse4_strcspn,0xa4e110);bl=0xa4e17e;addr=0xa4e17e;;xed decode error;no instruction at address
fn=(__intel_sse4_strcspn,0xa4e110);bl=0xa4e19a;addr=0xa4e19e;;xed decode error;no instruction at address
fn=(__intel_sse4_strcspn,0xa4e110);bl=0xa4e365;addr=0xa4e365;pcmpistrm;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strtok,0xa4e400);bl=0xa4e4a5;addr=0xa4e4a5;;xed decode error;no instruction at address
fn=(__intel_sse4_strtok,0xa4e400);bl=0xa4e679;addr=0xa4e679;pcmpistrm;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strcpy,0xa4e860);bl=0xa4e8e2;addr=0xa4e8e7;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strlen,0xa4e970);bl=0xa4e970;addr=0xa4e978;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strchr,0xa4e990);bl=0xa4e997;addr=0xa4e99b;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strend,0xa4e9c0);bl=0xa4e9c0;addr=0xa4e9c8;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strncmp,0xa4e9e0);bl=0xa4eb61;addr=0xa4eb66;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strcat,0xa4ecd0);bl=0xa4ecfa;addr=0xa4ed0b;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strcat,0xa4ecd0);bl=0xa4edc9;addr=0xa4edcd;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strncat,0xa4ef20);bl=0xa4ef50;addr=0xa4ef5f;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strncat,0xa4ef20);bl=0xa4f051;addr=0xa4f055;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_strdup,0xa4f210);bl=0xa4f27e;addr=0xa4f28e;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_atol,0xa4f370);bl=0xa4f551;addr=0xa4f551;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_new_memcpy,0xa52270);bl=0xa523d5;addr=0xa523d5;;xed decode error;no instruction at address
fn=(__intel_new_memcpy,0xa52270);bl=0xa523ee;addr=0xa523ee;;xed decode error;no instruction at address
fn=(_longjmp_unwind,0xa66950);bl=0xa6695a;addr=0xa6695d;jmp;(5 bytes);(xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(__libc_message,0xa8d770);bl=0xa8d7e9;addr=0xa8d7f3;;xed decode error;no instruction at address
fn=(__libc_message,0xa8d770);bl=0xa8d7ea;addr=0xa8d7f3;;xed decode error;no instruction at address
fn=(__strchr_sse42,0xa9a330);bl=0xa9a398;addr=0xa9a398;pcmpistri;(7 bytes);(xed length=7, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(__strcmp_sse42,0xa9b860);bl=0xa9b917;addr=0xa9b935;pcmpistri;(7 bytes);(xed length=7, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__rawmemchr_sse42,0xaa3850);bl=0xaa388a;addr=0xaa388a;pcmpestri;(7 bytes);(xed length=7, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__strend_sse4,0xaa6dc0);bl=0xaa6de3;addr=0xaa6de3;pcmpistri;(7 bytes);(xed length=7, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__strcspn_sse42,0xaa9e80);bl=0xaa9eb0;addr=0xaa9eb4;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__strcspn_sse42,0xaa9e80);bl=0xaa9ef8;addr=0xaa9ef8;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__strspn_sse42,0xaaa250);bl=0xaaa290;addr=0xaaa290;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__strspn_sse42,0xaaa250);bl=0xaaa318;addr=0xaaa31c;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__strstr_sse42,0xadcb60);bl=0xadcc08;addr=0xadcc19;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__strcasestr_sse42,0xade190);bl=0xade2b5;addr=0xade2c4;pcmpistri;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__m128i_strloadu_tolower_posix,0xade630);bl=0xade64d;addr=0xade661;pcmpistrm;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
lulesh-RAJA-parallel.exe.hpcstruct.report.txt
lulesh-RAJA-parallel.exe.report.txt
mhdshock.report.txt
pulse.report.txt
soliton.report.txt
ucns3d_p.report.txt
fn=(for_read_int_lis,0x4ad690);bl=0x4ad6eb;addr=0x4ad6f5;;xed decode error;no instruction at address
fn=(for_read_int_lis,0x4ad690);bl=0x4ad6ec;addr=0x4ad6f5;;xed decode error;no instruction at address
fn=(for_check_mult_overflow64,0x4c1870);bl=0x4c18d2;addr=0x4c18da;;xed decode error;no instruction at address
fn=(for_check_mult_overflow64,0x4c1870);bl=0x4c18d3;addr=0x4c18da;;xed decode error;no instruction at address
fn=(for_write_int_fmt,0x4c1aa0);bl=0x4c1af6;addr=0x4c1b00;;xed decode error;no instruction at address
fn=(for_write_int_fmt,0x4c1aa0);bl=0x4c1af7;addr=0x4c1b00;;xed decode error;no instruction at address
fn=(for_write_seq_lis,0x4d0600);bl=0x4d0656;addr=0x4d0660;;xed decode error;no instruction at address
fn=(for_write_seq_lis,0x4d0600);bl=0x4d0657;addr=0x4d0660;;xed decode error;no instruction at address
fn=(for_asynchronous,0x4db4a0);bl=0x4db4f9;addr=0x4db502;;xed decode error;no instruction at address
fn=(for_asynchronous,0x4db4a0);bl=0x4db4fa;addr=0x4db502;;xed decode error;no instruction at address
fn=(cvtas_a_to_x,0x50f980);bl=0x50f9d0;addr=0x50f9da;;xed decode error;no instruction at address
fn=(cvtas_a_to_x,0x50f980);bl=0x50f9d0;addr=0x50f9da;;xed decode error;no instruction at address
fn=(cvtas_a_to_x,0x50f980);bl=0x50f9d1;addr=0x50f9da;;xed decode error;no instruction at address
fn=(cvtas_a_to_x,0x50f980);bl=0x50f9d1;addr=0x50f9da;;xed decode error;no instruction at address
fn=(cos.N,0x51cb70);bl=0x51cb9c;addr=0x51cba8;roundsd;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(cos.N,0x51cb70);bl=0x51cf35;addr=0x51cf49;roundsd;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(sin.N,0x520570);bl=0x52059c;addr=0x5205b7;roundsd;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(sin.N,0x520570);bl=0x520949;addr=0x52095d;roundsd;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_memcpy,0x52fae0);bl=0x52fc45;addr=0x52fc45;;xed decode error;no instruction at address
fn=(__intel_memcpy,0x52fae0);bl=0x52fc5e;addr=0x52fc5e;;xed decode error;no instruction at address
xhpl.report.txt
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c23c;addr=0x46c23c;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c29e;addr=0x46c29e;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c2a2;addr=0x46c2a4;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c2a4;addr=0x46c2a4;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c2b6;addr=0x46c2b6;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c2ce;addr=0x46c2ce;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c2e5;addr=0x46c2e5;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c2e6;addr=0x46c2e6;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c2f3;addr=0x46c2f3;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c305;addr=0x46c305;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c319;addr=0x46c319;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c31a;addr=0x46c31a;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c324;addr=0x46c324;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c32a;addr=0x46c32a;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c32b;addr=0x46c32b;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c33a;addr=0x46c33b;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c35b;addr=0x46c35d;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down8x8zzai,0x46bf40);bl=0x46c360;addr=0x46c360;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46c9ed;addr=0x46c9ed;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46ca89;addr=0x46ca89;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46ca90;addr=0x46ca90;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cae7;addr=0x46cae7;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cb7a;addr=0x46cb7a;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cc09;addr=0x46cc09;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cc27;addr=0x46cc27;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cc2e;addr=0x46cc2e;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cd33;addr=0x46cd33;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cdb6;addr=0x46cdb6;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cdee;addr=0x46cdee;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cdee;addr=0x46cdee;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cf04;addr=0x46cf04;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cf62;addr=0x46cf62;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cfc1;addr=0x46cfc1;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cfc2;addr=0x46cfc2;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dcopy_down30x8zzoi,0x46c390);bl=0x46cfc8;addr=0x46cfc9;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_mic_dsymm_copy_upper_left_zz,0x7e3be0);bl=0x7e41f4;addr=0x7e42c4;vptestmd;(6 bytes);(xed length=6, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(mkl_blas_avx512_mic_dsymm_copy_lower_left_zz,0x7e4e20);bl=0x7e5498;addr=0x7e5567;vptestmd;(6 bytes);(xed length=6, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(mkl_blas_avx512_dgemm_ker0_pst,0x816580);bl=0x81005c;addr=0x810060;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_dgemm_ker0_pst,0x816580);bl=0x81005c;addr=0x810060;;xed decode error;no instruction at address
fn=(mkl_blas_avx512_dgemm_ker0_pst,0x816580);bl=0x813d3d;addr=0x813d3e;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad036;addr=0x9ad03e;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad036;addr=0x9ad03e;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad24e;addr=0x9ad24e;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad3f9;addr=0x9ad3f9;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad3f9;addr=0x9ad3f9;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad452;addr=0x9ad452;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad452;addr=0x9ad452;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad4a6;addr=0x9ad4a6;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad89e;addr=0x9ad89e;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ad8bb;addr=0x9ad8c9;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9adb32;addr=0x9adb32;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9adb5d;addr=0x9adb5d;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9adb5d;addr=0x9adb5d;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9adc51;addr=0x9adc51;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9adc6a;addr=0x9adc6a;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9adcd4;addr=0x9adcd7;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9add8a;addr=0x9add8a;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9addd2;addr=0x9addd3;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9addfa;addr=0x9addfa;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ade38;addr=0x9ade38;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ade95;addr=0x9ade97;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9adec7;addr=0x9adec7;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9adec7;addr=0x9adec7;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ae634;addr=0x9ae636;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ae637;addr=0x9ae63a;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9ae642;addr=0x9ae642;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9aeb51;addr=0x9aeb52;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9afcfd;addr=0x9afd07;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9afcfd;addr=0x9afd07;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b00b2;addr=0x9b00b2;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b00b2;addr=0x9b00b2;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b0573;addr=0x9b0573;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b05d7;addr=0x9b05df;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b05d7;addr=0x9b05df;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b0996;addr=0x9b0996;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b1b22;addr=0x9b1b22;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b1b22;addr=0x9b1b22;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b20a4;addr=0x9b20a4;pop;(2 bytes);(xed length=2, instruction api length=1) last instruction in block with DIRECT edge to sink
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b25ba;addr=0x9b25c2;;xed decode error;no instruction at address
fn=(MPID_Segment_manipulate,0x9aca30);bl=0x9b25ba;addr=0x9b25c2;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9adc51;addr=0x9adc51;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9adc6a;addr=0x9adc6a;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9adcd4;addr=0x9adcd7;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9add8a;addr=0x9add8a;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9addd2;addr=0x9addd3;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9addfa;addr=0x9addfa;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9ade38;addr=0x9ade38;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9ade95;addr=0x9ade97;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9adec7;addr=0x9adec7;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9adec7;addr=0x9adec7;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9ae634;addr=0x9ae636;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9ae637;addr=0x9ae63a;;xed decode error;no instruction at address
fn=(MPID_Segment_vector_m2m,0x9adc10);bl=0x9ae642;addr=0x9ae642;;xed decode error;no instruction at address
fn=(__I_MPI___intel_sse2_strtok,0xb1ada0);bl=0xb1ae77;addr=0xb1ae77;;xed decode error;no instruction at address
fn=(__I_MPI___intel_sse2_strtok,0xb1ada0);bl=0xb1ae93;addr=0xb1ae97;;xed decode error;no instruction at address
fn=(__I_MPI___intel_sse2_strtok,0xb1ada0);bl=0xb1b297;addr=0xb1b297;;xed decode error;no instruction at address
fn=(__I_MPI___intel_sse2_strtok,0xb1ada0);bl=0xb1b2a3;addr=0xb1b2a3;;xed decode error;no instruction at address
fn=(__I_MPI___intel_sse2_strtok,0xb1ada0);bl=0xb1b2af;addr=0xb1b2af;;xed decode error;no instruction at address
fn=(__I_MPI___intel_sse2_strtok,0xb1ada0);bl=0xb1b2bb;addr=0xb1b2bb;;xed decode error;no instruction at address
fn=(__I_MPI___intel_sse2_strtok,0xb1ada0);bl=0xb1b2d3;addr=0xb1b2d3;;xed decode error;no instruction at address
fn=(__I_MPI___intel_memcpy,0xb1c610);bl=0xb1c775;addr=0xb1c775;;xed decode error;no instruction at address
fn=(__I_MPI___intel_memcpy,0xb1c610);bl=0xb1c78e;addr=0xb1c78e;;xed decode error;no instruction at address
fn=(__intel_memcpy,0xb21c90);bl=0xb21df5;addr=0xb21df5;;xed decode error;no instruction at address
fn=(__intel_memcpy,0xb21c90);bl=0xb21e0e;addr=0xb21e0e;;xed decode error;no instruction at address

```

**Comment by wrwilliams** [comment]
Looks like the matching errors in that output (XED and IAPI agree there's no instruction) should mostly be over approximated jump tables or other extra blocks; does that match up with the objdump/manual analysis view of the world?

**Comment by mxz297** [comment]
Until now, my fixes should eliminated all over-approximation or wrong calculation of jump tables. However, after I pull current master to my local branch, I observed new instruction decoding problems, where several move instructions that load the table input were decoded as arpl instruction. I am able to reproduce this decoding error with the VEX branch, so I think John's previous fixes accidentally introduced this new bug. I have emailed John with more details to locate this bug.

**Comment by jdetter** [comment]
I'm still working on the instruction Xiaozhu sent me. It's an instruction in the `oneByteMap` that  has a completely different meaning if it's `0x66` prefixed. We don't have a way to differentiate these instructions from the `oneByteMap`. We do this in the other tables with the sseMap, sseMapBis and sseMapTer tables that come after the twoByteMap, threeByteMap and threeByteMap2.

@wrwilliams any suggestions on how we should handle this? Seems like an unusual case and I'm not sure if it's worth putting an instruction in a new table by itself.

**Comment by mxz297** [comment]
I opened a new issue with more details about this problem #88.

**Comment by jdetter** [comment]
I have created a new branch for the fixes for these issues, `release9.2/fixes/mellor-crummey-binaries`. The branch will be merged and deleted when the issues are fixed (hopefully in time for 9.2.X).

**Comment by jdetter** [comment]
It appears all of the instructions are now fixed on my end, however we are having a new issue on VEX that I'm looking into. #89 Hopefully this issue will be fixed by today so that you can give this another try.

**Comment by mxz297** [comment]
@jmellorcrummey We have pushed various fixes that should clear most of your error reportings. Could you test it soon to see how it works? Your feedback is important as we try to determine whether these fixes should be included in Dyninst 9.2 release.

**Comment by jmellorcrummey** [comment]
what branch has the fixes you would like me to evaluate?

**Comment by mxz297** [comment]
VEX. I am sorry that I forgot to mention this.

**Comment by jmellorcrummey** [comment]
The new result is much better! My tests still show a few issues with hpgmg-fv:

```
fn=(std::__enable_shared_from_this_helper<(__gnu_cxx::_Lock_policy)2>,0x57a230);bl=0x57a27f;addr=0x57a27f;;xed decode error;no instruction at address
fn=(gasnetc_AMRequestLongM,0x59df40);bl=0x59df80;addr=0x59df80;;xed decode error;no instruction at address
fn=(gasnetc_AMRequestLongAsyncM,0x59e310);bl=0x59e350;addr=0x59e350;;xed decode error;no instruction at address
fn=(gasnetc_AMReplyShortM,0x59e650);bl=0x59e69b;addr=0x59e69b;;xed decode error;no instruction at address
fn=(gasnetc_AMReplyMediumM,0x59e840);bl=0x59e885;addr=0x59e885;;xed decode error;no instruction at address
fn=(gasnetc_AMReplyLongM,0x59ea50);bl=0x59ea90;addr=0x59ea90;;xed decode error;no instruction at address
fn=(gasnetc_AMRequestMediumM,0x59ede0);bl=0x59ee23;addr=0x59ee23;;xed decode error;no instruction at address
fn=(gasnetc_AMRequestShortM,0x59f050);bl=0x59f099;addr=0x59f099;;xed decode error;no instruction at address
fn=(gasneti_fatalerror,0x5c6450);bl=0x5c64b0;addr=0x5c64b0;;xed decode error;no instruction at address
fn=(hp_timing_gettime,0x774b60);bl=0x774b8c;addr=0x774ba6;jmp;(5 bytes);(xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(MPL_cray_trdmapp_sheap_free,0x887420);bl=0x8874a1;addr=0x8874b0;jmp;(5 bytes);(xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(__svml_urem4_y8,0x9af5f0);bl=0x9af6e6;addr=0x9af6ea;roundpd;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_atol,0xa4f370);bl=0xa4f6ae;addr=0xa4f6ae;pextrd;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_atol,0xa4f370);bl=0xa4f6b9;addr=0xa4f6c9;pextrq;(7 bytes);(xed length=7, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(__intel_sse4_atol,0xa4f370);bl=0xa4f9c2;addr=0xa4f9f6;pextrd;(6 bytes);(xed length=6, instruction api length=4) last instruction in block with DIRECT edge to sink
fn=(_longjmp_unwind,0xa66950);bl=0xa6695a;addr=0xa6695d;jmp;(5 bytes);(xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(__rawmemchr_sse42,0xaa3850);bl=0xaa388a;addr=0xaa388a;pcmpestri;(7 bytes);(xed length=7, instruction api length=4) last instruction in block with DIRECT edge to sink
```

**Comment by jdetter** [comment]
I don't see an issue with the `jmp` instructions as far as decoding goes, @mxz297 would those be an issue on your end?

**Comment by mxz297** [comment]
I will take a look at this too. You @jdetter can start with address `a4f6ae`, which you can see through objdump.

**Comment by jdetter** [comment]
I just pushed fixes to VEX for the problem SSE instructions  e71593fde90d2b0e8e46411c9d765747d6ce4b3f

**Comment by jmellorcrummey** [comment]
That led to a build failure for me:

/home/johnmc/Documents/pkgs-src/dyninst/common/src/arch-x86.C:5639:63: error: 'Eq' was not declared in this scope
         { e_pextrd_pextrq, t_sse_ter_mult, SSET16_66, true, { Eq, Vdq, Ib }, 0, s1W2R3R },

**Comment by jdetter** [comment]
Yup that's what I get for not rebuilding, sorry about that! 36911135c1baeef12b1a79c4a38d095543ddf06a will fix the build issue.

**Comment by jdetter** [comment]
@mxz297 is taking a look at why the jump instructions aren't behaving nicely.

**Comment by mxz297** [comment]
@jmellorcrummey For the following two reporting
```

fn=(hp_timing_gettime,0x774b60);bl=0x774b8c;addr=0x774ba6;jmp;(5 bytes);(xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
fn=(MPL_cray_trdmapp_sheap_free,0x887420);bl=0x8874a1;addr=0x8874b0;jmp;(5 bytes);(xed length=5, instruction api length=5) last instruction in block with DIRECT edge to sink
```

Dyninst shows that they are `jmp 0` and Objdump shows that they are `jmpq   0 <USE_REG>` too. So, in these cases, we are reporting what the binary says... 

I am continuing on other reports.

**Comment by jmellorcrummey** [comment]
I don't dispute the decoding on these two instructions. I am concerned about the CFG though. For the first, the snippet around the instruction of interest is:

```
/usr/src/packages/BUILD/glibc-2.11.3/rt/../sysdeps/unix/clock_gettime.c:75
  774b98:       48 8b 5c 24 08          mov    0x8(%rsp),%rbx
  774b9d:       48 8b 6c 24 10          mov    0x10(%rsp),%rbp
  774ba2:       48 83 c4 18             add    $0x18,%rsp
/usr/src/packages/BUILD/glibc-2.11.3/rt/../sysdeps/unix/clock_gettime.c:59
  774ba6:       e9 55 b4 88 ff          jmpq   0 <USE_REG>
  774bab:       0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
/usr/src/packages/BUILD/glibc-2.11.3/rt/../sysdeps/unix/clock_gettime.c:62
  774bb0:       0f 31                   rdtsc
  774bb2:       89 c1                   mov    %eax,%ecx
  774bb4:       48 89 d0                mov    %rdx,%rax
/usr/src/packages/BUILD/glibc-2.11.3/rt/../sysdeps/unix/clock_gettime.c:68
  774bb7:       48 8b 35 c2 86 ad 00    mov    0xad86c2(%rip),%rsi        # 124d280 <freq>
/usr/src/packages/BUILD/glibc-2.11.3/rt/../sysdeps/unix/clock_gettime.c:62
  774bbe:       89 ca                   mov    %ecx,%edx

```

My debugging harness shows the jmpq 0 to be the last instruction in a basic block with a DIRECT edge to sink. Isn't "jmpq 0" the equivalent of a control flow no-op? As I read the Intel instruction set manual, a jump with a relative offset jumps to RIP + offset, where the RIP points to the instruction after the jump.

**Comment by mxz297** [comment]
I see. Let me explain this a little bit more. When I say `jmp 0` from Dyninst or Objdump, the 0 is not the relative offset of the jump. The 0 is the calculation result of RIP+ offset. So, if this instruction is executed, the control flow will go to virtual address 0 and segfault. That's why Dyninst says it goes to sink.

**Comment by jmellorcrummey** [comment]
I studied the situation a bit more and I agree. This is sure a strange instruction to have in the middle of a procedure!

**Comment by jdetter** [comment]
Bart is deciding whether or not these fixes are going into the 9.2.X release. The faster we get these cleaned up the more likely it will be for these changes to make it in. I think all of us would really like to see these fixes get into 9.2.X.

I sent mail offline so Bart can see the progress on this issue.

If there is anything left that I can help with please let me know right away. I will be watching my email and this thread very closely today and tomorrow, so if anything comes up I will get on it right away. For now, I will be working on getting #90 resolved as quickly as possible.

**Comment by mxz297** [comment]
@jmellorcrummey Finally I pushed extra fixes that should eliminate all error reportings. All fixes are on VEX branch. 

We spent some time on the `jmp 0`. By comparing the binary code and the source code, we think the jump should be a tail call to `__pthread_clock_gettime`. And we find that the symbol table says:

`25745: 0000000000000000     0 NOTYPE  WEAK   DEFAULT  UND __pthread_clock_gettime`

So,  the binary contains tail calls to undefined symbols ...

**Comment by jmellorcrummey** [comment]
Your explanation makes a lot of sense. It seems a miracle that the linker didn't balk!

**Comment by jmellorcrummey** [comment]
Everything looks good on the VEX branch relative to my test cases. You guys made enormous progress over the last week. Thanks for thoroughly addressing the issues that I reported!

**Comment by wrwilliams** [comment]
The linker didn't balk because it's a weak symbol, but sheesh. This code is indeed stupid.

**Comment by mxz297** [comment]
Great to know that all your test cases are good!
